### DISCUSSION TITLE
Why don't you use Elasticsearch instead of PostgreSQL

### DISCUSSION THREAD

**Discussion by stouch** [osm-search/Nominatim#3004] [Ideas]
Is there any specific reason for which you index and request the data in a postgreSQL db and not in a document-based such ES ?

Would it not be faster in indexing and requesting ?

Regards

**Comment by lonvia** [comment]
There is an ES frontend. See https://github.com/komoot/photon/.

**Comment by stouch** [comment]
I mean, why dont you upgrade this project to a ES index mechanism instead of postgresql ? Just to avoid the migration for users, or is there any perf reason to keep it on postgresql ?

**Reply by mtmail** [reply]
Twice you touch the subject of performance. Is Nominatim not performant enough for you use case? In case of request performance you can try Photon, it used Nominatim data. With so many open issues, a long wishlist of features and few resources (people) why embark on a risky possibly (multi?)year rewrite. If a ES based geocoder comes along and is much superior users will switch, it's the nature of competition in the open source world and having multiple striving open source geocoders (separate teams, different motivations and focus) is good for all users.

**Reply by stouch** [reply]
The perf question is to understand the reason the team kept this db instead of another one, And also because the importing process is very long so maybe it'd be faster with ES for example (or redis like Addock does in France : https://github.com/addok/addok

**Reply by mtmail** [reply]
Addok imports 4 million lines like
```
{"id":"01012","type":"village", "name":"Aranc", "postcode":"01110", "lat":46.003580, "lon":5.507669, "city":"Aranc", "departement":"Ain", "region":"Auvergne-Rhône-Alpes", "population":326, "adm_weight":1, "importance":0.2644}
{"id":"01013","type":"village", "name":"Arandas", "postcode":"01230", "lat":45.896039, "lon":5.487706, "city":"Arandas", "departement":"Ain", "region":"Auvergne-Rhône-Alpes", "population":144, "adm_weight":1, "importance":0.2356}
{"id":"01014","type":"town", "name":"Arbent", "postcode":"01100", "lat":46.294668, "lon":5.680467, "city":"Arbent", "departement":"Ain", "region":"Auvergne-Rhône-Alpes", "population":3367, "adm_weight":1, "importance":0.3376}
```
Redis is a good choice for that project. Their data store is about 4GB, usually fits into RAM and thus very fast. The input data is from a government source, very clean and defined and likely complete.

Nominatim plays in another league in terms of features, complexity of input data and data size. It can update incremental after OpenStreetMap data changes, deal with border polygons, address interpolation, places having names in multiple languages (plus short, long, official names), different address hierarchies and missing data. Postgresql+PostGIS has been a good choice so far and continues to be. I think somebody could rewrite Nominatim using a different database but it's too time consuming, risky and I personally don't see benefits.

Import is slow indeed. Usually you'd only need to run it once and then with minutely/daily/monthly updates never again.

**Reply by stouch** [reply]
Thank you  😊 (Indeed, they work this kind of data thanks to a french opensource initiative with the collaboration of many cities in France https://www.data.gouv.fr/fr/datasets/base-d-adresses-nationale-ouverte-bano/)
